### PR TITLE
torchdynamo and xla integration

### PR DESCRIPTION
**Pull request by shunting314**
# Motivation
- torchdynamo and torchxla uses different strategies to be a sound graph capture technique. The former relies on guards; the latter relies on retracing
- guard system is quite low overhead but torchxla tracing overhead is quite high

The main idea is to leverage guard system in torchdynamo to avoid retracing in torchxla so that
- we can integration torchdynamo with XLA
- we reduce or even completely avoid tracing overhead of torchxla

# Technique details
## XLA baseline
We found that different frameworks do not generate numerically identical results for the SAME model with the SAME input. By default, torchdynamo uses eager as baseline so the model will run with PyTorch. It would be tricky to compare a model running on XLA with this baseline: it's hard to check correctness. To make the comparison easier, we add a flag `--use-xla-baseline`. When it's enabled, the baseline will be run on XLA.

## New dynamo backends added
We add 2 new dynamo backends torchxla_trivial and trochxla_trace_once to control the optimization targets.

torchxla_trivial simply moves inputs/model parameters to XLA and run the model on XLA. There is tracing overhead for each run. We should expect that result to be mostly neutral compared to the XLA baseline.

torchxla_trace_once only traces once during AOT compiling time. Here are the steps:
1. dynamo capture guards and the subgraph
2. torchxla_trace_once backend trace the graph with torchxla, lowering the graph and record a hash of the graph for later lookup
3. at inference time, the hash is used directly to lookup the optimized graph and run it.

# Limitations
We can not handle LTC/torchxla fall back right now. If a op misses LTC kernel, we raise and exception and that will results in dynamo fallback (or try another compiler). People have brainstormed the idea of graph breaking and stitching the subgraphs together. But maybe it's easier to add those missing LTC kernels for those models.

# Results
The models we tested are those not causing LTC fallback. We run the tests on **GPU**. We see **1.38x** geomean speedup for trochxla_trace_once  and torchxla_trivial is mostly neutral as expected.
```
| Model                   |   XLA (trace once) |   XLA (trace everytime) |
+=========================+====================+=========================+
| resnet18                |            1.346   |                 1.045   |
+-------------------------+--------------------+-------------------------+
| resnet50                |            1.153   |                 1.007   |
+-------------------------+--------------------+-------------------------+
| resnext50_32x4d         |            1.381   |                 1.039   |
+-------------------------+--------------------+-------------------------+
| alexnet                 |            1.045   |                 1.018   |
+-------------------------+--------------------+-------------------------+
| mobilenet_v2            |            1.562   |                 1.021   |
+-------------------------+--------------------+-------------------------+
| mnasnet1_0              |            1.303   |                 1.069   |
+-------------------------+--------------------+-------------------------+
| squeezenet1_1           |            1.278   |                 1.025   |
+-------------------------+--------------------+-------------------------+
| vgg16                   |            1.076   |                 1.008   |
+-------------------------+--------------------+-------------------------+
| BERT_pytorch            |            2.224   |                 0.978   |
+-------------------------+--------------------+-------------------------+
| timm_vision_transformer |            1.81    |                 1.025   |
+-------------------------+--------------------+-------------------------+
| geomean                 |            1.38101 |                 1.02324 |
+-------------------------+--------------------+-------------------------+
```

The speedup is similar to what we see from previous work for LTC's TorchScript backend (we see 1.40 geomean speedup there):
https://docs.google.com/presentation/d/1G09X8v41u_cLKLtSdf7v6R8G19-iZTPcW_VAdOnvYBI/edit#slide=id.g11bf989cb6b_1_5

# Next steps
- Use AOT autograd to enable training
- Share results on XLA devices
- Do more extensive tests on torchbench models

Example command
```
GPU_NUM_DEVICES=1 python benchmarks/dynamo/torchbench.py --randomize-input --performance --use-xla-baseline --only resnet18 --backend=torchxla_trace_once
```

Thanks @JackCaoG from torchxla team to help debugging various perf issues and merging the torchxla PR! That's super critical for us to get the results above. torchxla side PR: https://github.com/pytorch/xla/pull/4119 

topic: not user facing

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @jansel

### CODE REVIEW

**Review by wconstab — APPROVED** [review]
looks good, but i do think it is worth reconsidering how the UX of moving tensors to XLA device is implemented.  I'm ok landing it and then updating if that unblocks XLA team in the mean-time.

**Comment by JackCaoG** [comment]
TPU v4-8 single device speed up is
| model name        | speed up           | lazy time  | dynamo time
| ------------- |:-------------:| -----:|-----:
resnet50 | 1.112 | 0.01545586 | 0.01389865
resnext50_32x4d | 1.424 | 0.01714463 | 0.01204363
alexnet | 0.697 | 0.0620243 | 0.08900219
mobilenet_v2 | 1.505 | 0.01111322 | 0.00738665
mnasnet1_0 | 1.114 | 0.0136932 | 0.01229267
squeezenet1_1 | 0.825 | 0.00689901 | 0.0083646
vgg16 | 0.994 | 0.00598695 | 0.00602498
BERT_pytorch | 3.466 | 0.01493634 | 0.00430994
timm_vision_transformer | 2.611 | 0.01517026 | 0.00580959

on bert and vit the speed up is very significant. It is puzzling why alexnet and squeezenet1_1 is slower with dynamo.

File: torch/_dynamo/optimizations/torchxla_integration/__init__.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
+import copy
```
Comment (jansel) [inline torch/_dynamo/optimizations/torchxla_integration/__init__.py:1]: I'd suggest just putting this file under `torch/_dynamo/optimizations/torchxla_integration.py`, unless you expect it to grow large enough to be many files.

File: torch/_dynamo/optimizations/torchxla_integration/tests/reuse_graph_test.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
+import copy
```
Comment (jansel) [inline torch/_dynamo/optimizations/torchxla_integration/tests/reuse_graph_test.py:1]: Let's put test files in the top level `test/dynamo` directory, with the right skips if XLA isn't installed.

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by shunting314** [comment]
> TPU v4-8 single device speed up is
> 
> model name	speed up	lazy time	dynamo time
> resnet50	1.112	0.01545586	0.01389865
> resnext50_32x4d	1.424	0.01714463	0.01204363
> alexnet	0.697	0.0620243	0.08900219
> mobilenet_v2	1.505	0.01111322	0.00738665
> mnasnet1_0	1.114	0.0136932	0.01229267
> squeezenet1_1	0.825	0.00689901	0.0083646
> vgg16	0.994	0.00598695	0.00602498
> BERT_pytorch	3.466	0.01493634	0.00430994
> timm_vision_transformer	2.611	0.01517026	0.00580959
> on bert and vit the speed up is very significant. It is puzzling why alexnet and squeezenet1_1 is slower with dynamo.

Try `--torchxla-trivial` on these 2 models? `--torchxla-trivial` enables dynamo but do no avoid retracing. That can be used to measure dynamo overhead here.

File: torch/_dynamo/optimizations/torchxla_integration/__init__.py (hunk 1 above)
Comment (shunting314) [reply torch/_dynamo/optimizations/torchxla_integration/__init__.py:1]: I can move it. The file size should still be manageable in near future.

File: benchmarks/dynamo/common.py [hunk 3]
```diff
@@ -783,6 +786,51 @@ def inner(self, *args, **kwargs):
...
+
+    def wrapper(mod, inputs, collect_outputs=True):
+        import torch_xla.core.xla_model as xm
+
+        xla_dev = xm.xla_device()
+        eager_dev = inputs[0].device
+        # We assume the passed in mod is already on xla device
+        # so we dont need: xla_mod = copy.deepcopy(mod).to(device=xla_dev)
+        xla_mod = mod
+        xla_inputs = tree_map(lambda x: x.to(device=xla_dev), inputs)
```
[the remark this answers is not in the document]
Comment (shunting314) [reply benchmarks/dynamo/common.py:802]: yes, `inputs.to(lazy_device)` and `outputs.to(eager_device)` are included in the timing for baseline. Those overhead are included in the timing for `torchxla_trace_once` backend as well.

File: torch/_dynamo/optimizations/backends.py [hunk 4]
```diff
@@ -785,6 +785,44 @@ def ltc_model(*inputs):
...
+def torchxla_trivial(subgraph):
+    _init_torchxla()
+
+    xla_dev = xm.xla_device()
+
+    xla_model = copy.deepcopy(subgraph.model).to(device=xla_dev)
+
+    def xla_model_wrapper(*inputs):
+        orig_device = inputs[0].device if len(inputs) > 0 else "cpu"
+        xla_inputs = tuple(inp.to(device=xla_dev) for inp in inputs)
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_dynamo/optimizations/backends.py:808]: So basically you mean wrap the `model_iter_fn` once (to handle the cross device movements) so we don't need to do it in any other places?

I actually tried this idea since I agree that makes the code cleaner. But the problem I encountered is, the wrapped `model_iter_fn` fail to be compiled by dynamo. I guess there is some incompatibility if we apply XLA and dynamo (graph capture part rather than the backend) together.

File: torch/_dynamo/optimizations/torchxla_integration/__init__.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        print(f"XLA IR HLO: {torch_xla._XLAC._get_xla_tensors_hlo(args_and_out)}")
+
+    # calculate graph hash
+    graph_hash = torch_xla._XLAC._get_graph_hash(args_and_out)
+    if debug:
+        print("graph_hash", graph_hash)
+
+    (
+        graph_input_tensor_ids,
+        graph_input_ivalues,
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_dynamo/optimizations/torchxla_integration/__init__.py:112]: haha, call it `ivalue` since we return IValue in the pybind API:
```
  m.def("_get_tensors_xla_device_data_node",
        [](const std::vector<at::Tensor>& tensors)
            -> std::pair<std::vector<int64_t>, std::vector<at::IValue>> {
 ```

[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_dynamo/optimizations/torchxla_integration/__init__.py:112]: `backend_handles` would be more suitable for some opaque lookup keys which will be interpreted somewhere else (i.e. in XLA)?

Just double check. I'm ok to rename though.

**Comment by shunting314** [comment]
Resolve comments as much as I can. Some left ones need some clarification.

@jansel, @wconstab   do you want to take another look

**Comment by shunting314** [comment]
The failed job '
trunk / linux-bionic-cuda11.7-py3.10-gcc7 / test (nogpu_NO_AVX2, 1, 1, linux.2xlarge) (push) ' has a very long log. Here is the most relevant error message snippet I can find:
```

2022-10-29T03:03:55.3797957Z [0;32m[ RUN      ] [mQuantization.QuantUpsampleNearst2dDequantUInt8
2022-10-29T03:03:55.4259856Z x:
2022-10-29T03:03:55.4261226Z (1,1,.,.) = 
2022-10-29T03:03:55.4261716Z   0.1066  0.6779  0.1332  0.2720
2022-10-29T03:03:55.4261979Z   0.7040  0.0439  0.8677  0.0901
2022-10-29T03:03:55.4262195Z   0.2853  0.6457  0.2500  0.6500
2022-10-29T03:03:55.4262449Z   0.7511  0.3352  0.9910  0.4533
2022-10-29T03:03:55.4262860Z [ CPUFloatType{1,1,4,4} ]
2022-10-29T03:03:55.4263085Z q:
2022-10-29T03:03:55.4263239Z (1,1,.,.) = 
2022-10-29T03:03:55.4263424Z   0.1000  0.7000  0.1000  0.3000
2022-10-29T03:03:55.4263582Z   0.7000  0.0000  0.9000  0.1000
2022-10-29T03:03:55.4263817Z   0.3000  0.6000  0.3000  0.6000
2022-10-29T03:03:55.4263996Z   0.8000  0.3000  1.0000  0.5000
2022-10-29T03:03:55.4264338Z [ QuantizedCPUQUInt8Type{1,1,4,4}, qscheme: per_tensor_affine, scale: 0.1000, zero_point: 13 ]
2022-10-29T03:03:55.4264632Z qu:
2022-10-29T03:03:55.4264786Z (1,1,.,.) = 
2022-10-29T03:03:55.4264966Z   0.1000  0.1000  0.7000  0.1000  0.1000  0.3000
2022-10-29T03:03:55.4265145Z   0.1000  0.1000  0.7000  0.1000  0.1000  0.3000
2022-10-29T03:03:55.4265389Z   0.7000  0.7000  0.0000  0.9000  0.9000  0.1000
2022-10-29T03:03:55.4265588Z   0.3000  0.3000  0.6000  0.3000  0.3000  0.6000
2022-10-29T03:03:55.4265764Z   0.3000  0.3000  0.6000  0.3000  0.3000  0.6000
2022-10-29T03:03:55.4265952Z   0.8000  0.8000  0.3000  1.0000  1.0000  0.5000
2022-10-29T03:03:55.4266283Z [ QuantizedCPUQUInt8Type{1,1,6,6}, qscheme: per_tensor_affine, scale: 0.1000, zero_point: 13 ]
2022-10-29T03:03:55.4266523Z y_expected:
2022-10-29T03:03:55.4266686Z (1,1,.,.) = 
2022-10-29T03:03:55.4266904Z   0.1000  0.1000  0.7000  0.1000  0.1000  0.3000
2022-10-29T03:03:55.4267090Z   0.1000  0.1000  0.7000  0.1000  0.1000  0.3000
2022-10-29T03:03:55.4267278Z   0.7000  0.7000  0.0000  0.9000  0.9000  0.1000
2022-10-29T03:03:55.4267467Z   0.3000  0.3000  0.6000  0.3000  0.3000  0.6000
2022-10-29T03:03:55.4267701Z   0.3000  0.3000  0.6000  0.3000  0.3000  0.6000
2022-10-29T03:03:55.4267883Z   0.8000  0.8000  0.3000  1.0000  1.0000  0.5000
2022-10-29T03:03:55.4268076Z [ CPUFloatType{1,1,6,6} ]
2022-10-29T03:03:55.4268253Z y:
2022-10-29T03:03:55.4268412Z (1,1,.,.) = 
2022-10-29T03:03:55.4268624Z   0.1000  0.1000  0.7000  0.1000  0.1000  0.3000
2022-10-29T03:03:55.4268812Z   0.1000  0.1000  0.7000  0.1000  0.1000  0.3000
2022-10-29T03:03:55.4268982Z   0.7000  0.7000  0.0000  0.9000  0.9000  0.1000
2022-10-29T03:03:55.4269186Z   0.3000  0.3000  0.6000  0.3000  0.3000  0.7000
2022-10-29T03:03:55.4269410Z   0.3000  0.3000  0.6000  0.3000  0.3000  0.7000
2022-10-29T03:03:55.4269582Z   0.8000  0.8000  0.3000  1.0000  1.0000  0.5000
2022-10-29T03:03:55.4269767Z [ CPUFloatType{1,1,6,6} ]
2022-10-29T03:03:55.4270073Z [F test_quantization.cpp:352] Check failed: check == 1 (0 vs. 1) 
2022-10-29T03:03:55.5846765Z .jenkins/pytorch/test.sh: line 355: 14643 Aborted                 (core dumped) "$TORCH_BIN_DIR"/test_tensorexpr --gtest_output=xml:$TEST_REPORTS_DIR/test_tensorexpr.xml 
```

Looks like some TORCH_CHECK fail for quantization and cause the testing process abort.

I think it's not related to the PR.

File: benchmarks/dynamo/common.py [hunk 6]
```diff
@@ -784,6 +787,56 @@ def inner(self, *args, **kwargs):
...
+            next(xla_mod.parameters()).device.type == "xla"
+        ), "The model should be already on xla device"
+
+        xla_dev = xm.xla_device()
+        eager_dev = inputs[0].device
+        xla_inputs = tree_map(lambda x: x.to(device=xla_dev), inputs)
+        xla_out = model_iter_fn(xla_mod, xla_inputs, collect_outputs)
+        if isinstance(xla_out, torch.Tensor):
+            return xla_out.to(device=eager_dev)
+        elif hasattr(xla_out, "__dict__"):
```
Comment (byronyi) [inline benchmarks/dynamo/common.py:812]: If this is for hf transformer, you could register its output type to pytree like in https://github.com/pytorch/functorch/issues/475#issuecomment-1035983774.

File: torch/_dynamo/optimizations/backends.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (byronyi) [reply torch/_dynamo/optimizations/backends.py:808]: Since you are targeting on training later on, there is actually a head-up: random functions, e.g. torch.aten.ops.dropout, will require an extra seed input in torch-xla lowering. This seed, unfortunately, is not a PyTorch tensor input in the fx Graph, so there is no way to pass it with the `inp.to(device=xla)`. PyTorch pretends dropout are stateless and functional while it is really not.

This issue also applies to other non-Tensor scalars, but that is more of an issue for dynamo.
